ts/src/base/Exchange.ts

```diff
@@ -4541,6 +4541,206 @@ export default class Exchange {
+                    }
+                } else {
+                    // do it forwards, starting from the since
+                    const response = await this[method] (symbol, paginationTimestamp, maxEntriesPerRequest, params);
+                    const responseLength = response.length;
+                    if (this.verbose) {
+                        this.log ('Dynamic pagination call', calls, 'method', method, 'response length', responseLength, 'timestamp', paginationTimestamp);
+                    }
+                    if (responseLength === 0) {
+                        break;
+                    }
+                    errors = 0;
+                    result = this.arrayConcat (result, response);
+                    const last = this.safeValue (response, responseLength - 1);
+                    paginationTimestamp = this.safeInteger (last, 'timestamp') - 1;
+                }
+            } catch (e) {
+                errors += 1;
+                if (errors > maxRetries) {
+                    throw e;
+                }
+            }
+        }
+        return this.removeRepeatedElementsFromArray (result);
+    }
+
+    async safeDeterministicCall (method, symbol = undefined, since = undefined, limit = undefined, timeframe = undefined, params = {}) {
+        let maxRetries = undefined;
+        [ maxRetries, params ] = this.handleOptionAndParams (params, method, 'maxRetries', 3);
+        let errors = 0;
+        try {
+            if (timeframe && method !== 'fetchFundingRateHistory') {
+                return await this[method] (symbol, timeframe, since, limit, params);
+            } else {
+                return await this[method] (symbol, since, limit, params);
+            }
+        } catch (e) {
+            errors += 1;
+            if (errors > maxRetries) {
+                throw e;
+            }
+        }
+    }
+
+    async fetchPaginatedCallDeterministic (method: string, symbol: string = undefined, since = undefined, limit = undefined, timeframe = undefined, params = {}, maxEntriesPerRequest = undefined): Promise<any> {
+        let maxCalls = undefined;
+        [ maxCalls, params ] = this.handleOptionAndParams (params, method, 'paginationCalls', 10);
+        [ maxEntriesPerRequest, params ] = this.handleMaxEntriesPerRequestAndParams (method, maxEntriesPerRequest, params);
+        const current = this.milliseconds ();
+        const tasks = [];
+        const time = this.parseTimeframe (timeframe) * 1000;
+        const step = time * maxEntriesPerRequest;
+        let currentSince = current - (maxCalls * step) - 1;
+        if (since !== undefined) {
+            currentSince = Math.max (currentSince, since);
+        }
+        for (let i = 0; i < maxCalls; i++) {
+            tasks.push (this.safeDeterministicCall (method, symbol, currentSince, maxEntriesPerRequest, timeframe, params));
+            currentSince = currentSince + step - 1;
+        }
+        const results = await Promise.all (tasks);
```

Comment:
I did a very large call to create a rate limit error, and it stays calling and failing. Ideally I think in this case we would want to fail fast.
Or instead of using Promise.all if we have any other alternatives (like an await of), so on each fullfiled promise, if it's resolved its pushed to the results and if it's rejected we throw an exception.

Or if not would it make sense to check the results for possible rejected promises?

Reply:
> I did a very large call to create a rate limit error, and it stays calling and failing. Ideally I think in this case we would want to fail fast.

what if we checked the error type and if is a `RateLimit` error we throw instead of continuing?

> Or instead of using Promise.all if we have any other alternatives (like an await of), so on each fullfiled promise, if it's resolved its pushed to the results and if it's rejected we throw an exception.

sorry, not sure I get this, promise.all will throw if one of the promises is throws/rejects right?

Reply:
> what if we checked the error type and if is a RateLimit error we throw instead of continuing?
Yes, I think that solves it staying hanged
> sorry, not sure I get this, promise.all will throw if one of the promises is throws/rejects right?
You are right, I saw it get hanged and thought it was because of the promiseAll but it would get solved with the previous.

Reply:
> what if we checked the error type and if is a RateLimit error we throw instead of continuing?
Yes, I think that solves it staying hanged
> sorry, not sure I get this, promise.all will throw if one of the promises is throws/rejects right?
You are right, I saw it get hanged and thought it was because of the promiseAll but it would get solved with the previous.